src/Features/Core/Portable/PullMemberUp/MemberAndDestinationValidator.cs

```diff
@@ -0,0 +1,37 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.  
+
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.PullMemberUp
+{
+    internal static class MemberAndDestinationValidator
+    {
+        internal static bool IsDestinationValid(INamedTypeSymbol destination, Solution solution, CancellationToken cancellationToken)
+        {
+            return destination != null &&
+                destination.DeclaringSyntaxReferences.Length != 0 &&
+                (destination.TypeKind == TypeKind.Interface || destination.TypeKind == TypeKind.Class) &&
+                destination.Locations.Any(location => location.IsInSource &&
+                !solution.GetDocument(location.SourceTree).IsGeneratedCode(cancellationToken));
```

Comment:
this is all just confusing.  Can you ust break this into separate statements, commenting on why each case is appropriate?

Reply:
Sure